vochain/account.go

```diff
@@ -175,108 +214,510 @@ func (v *State) GetAccount(address common.Address, isQuery bool) (*Account, erro
+	returnValues := &setAccountInfoTxCheckValues{}
 	tx := vtx.GetSetAccountInfo()
 	// check signature available
 	if signature == nil || tx == nil || txBytes == nil {
-		return common.Address{}, false, fmt.Errorf("missing signature and/or transaction")
+		return nil, fmt.Errorf("missing signature and/or transaction")
 	}
+	// check infoURI
 	infoURI := tx.GetInfoURI()
 	if infoURI == "" {
-		return common.Address{}, false, fmt.Errorf("invalid URI, cannot be empty")
+		return nil, fmt.Errorf("invalid URI, cannot be empty")
+	}
+	// get tx cost
+	cost, err := state.TxCost(models.TxType_SET_ACCOUNT_INFO, false)
+	if err != nil {
+		return nil, err
+	}
+	// recover txSender address from signature
+	pubKey, err := ethereum.PubKeyFromSignature(txBytes, signature)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract public key from signature: %w", err)
+	}
+	returnValues.TxSender, err = ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	// get txSender account
+	txSender, err := state.GetAccount(returnValues.TxSender, false)
+	if err != nil {
+		return nil, fmt.Errorf("cannot check if account %s exists: %v", returnValues.TxSender.String(), err)
+	}
+	// get tx.Account, if not valid set to txSender address
+	accountAddressBytes := tx.GetAccount()
+	returnValues.Account = common.BytesToAddress(accountAddressBytes)
+	if accountAddressBytes == nil ||
+		len(accountAddressBytes) != types.EntityIDsize ||
+		bytes.Equal(accountAddressBytes, types.EthereumZeroAddressBytes[:]) {
+		returnValues.Account = returnValues.TxSender
+	}
+	// if not exist create new one
+	if txSender == nil {
+		if tx.FaucetPackage != nil {
+			if bytes.Equal(tx.FaucetPackage.Payload.To, returnValues.TxSender.Bytes()) {
+				// get issuer address from faucetPayload
+				faucetPkgPayload := tx.FaucetPackage.GetPayload()
+				faucetPackageBytes, err := proto.Marshal(faucetPkgPayload)
+				if err != nil {
+					return nil, fmt.Errorf("cannot extract faucet package payload: %v", err)
+				}
+				issuerAddress, err := ethereum.AddrFromSignature(faucetPackageBytes, tx.FaucetPackage.Signature)
+				if err != nil {
+					return nil, err
+				}
+				// check issuer nonce not used
+				b := make([]byte, 8)
+				binary.LittleEndian.PutUint64(b, faucetPkgPayload.Identifier)
+				key := issuerAddress.Bytes()
+				key = append(key, b...)
+				used, err := state.FaucetNonce(crypto.Sha256(key), false)
+				if err != nil {
+					return nil, fmt.Errorf("cannot check faucet nonce: %v", err)
+				}
+				if used {
+					return nil, fmt.Errorf("nonce %d already used", faucetPkgPayload.Identifier)
+				}
+				// check issuer have enough funds
+				issuerAcc, err := state.GetAccount(issuerAddress, false)
+				if err != nil {
+					return nil, fmt.Errorf("cannot get faucet account: %v", err)
+				}
+				if issuerAcc == nil {
+					return nil, ErrAccountNotFound
```

Comment:
wrap this to clarify the faucet account, not the recipient account, was not found